linker: arm: Fix and enhance Cortex-M TCM support

MaureenHelm opened the pull request:
Fixes the Cortex-M DTCM linker section to work with relocation using scripts/gen_relocate_app.py

Adds support for the Cortex-M ITCM linker section.

This is in preparation for extending #31034 to support XIP.

MaureenHelm committed:
linker: arm: Fix cortex_m dtcm and ccm sections to work with relocation

Adds missing DTCM_ADDR and CCM_ADDR macros needed by the linker script
generated by scripts/gen_relocate_app.py.

Moves associated bss_end, noinit_end, and data_end linker symbols to
account for section relocation. Without this change, the section sizes
calculated in z_bss_zero() and z_data_copy() are incorrect.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
linker: arm: Add cortex_m itcm section

Adds a linker section for Cortex-M instruction tightly coupled memory
(ITCM), similar to the existing section for DTCM. A new executable MPU
region is not added as there isn't currently a need to make this section
accessible to user mode. This section can be enabled by setting a device
tree chosen node zephyr,itcm.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

erwango commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -99,7 +104,7 @@ MEMORY
     CCM                   (rw) : ORIGIN = DT_REG_ADDR(DT_CHOSEN(zephyr_ccm)), LENGTH = DT_REG_SIZE(DT_CHOSEN(zephyr_ccm))
 #endif
 #if DT_NODE_HAS_STATUS(DT_CHOSEN(zephyr_dtcm), okay)
-    DTCM                  (rw) : ORIGIN = DT_REG_ADDR(DT_CHOSEN(zephyr_dtcm)), LENGTH = DT_REG_SIZE(DT_CHOSEN(zephyr_dtcm))
+    DTCM                  (rw) : ORIGIN = DTCM_ADDR, LENGTH = DTCM_SIZE
```
For coherency sake, I wonder if it would makes sense to do the same for ccm, so that ccm, itcm and dtcm are fully aligned following your change.

MaureenHelm replied:
Sure, done. I don't have hardware to test it though, will you please check it?

erwango replied:
Tested ok on nucleo_f429z (using samples/boards/olimex_stm32_e407/ccm)

ioannisg commented on arch/arm/core/aarch32/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -77,6 +77,14 @@ static const struct z_arm_mpu_partition static_regions[] = {
 		.attr = K_MEM_PARTITION_P_RX_U_RX,
 		},
 #endif /* CONFIG_ARCH_HAS_RAMFUNC_SUPPORT */
+#if DT_NODE_HAS_STATUS(DT_CHOSEN(zephyr_itcm), okay)
```
@MaureenHelm this will consume one MPU region.
I wonder whether this is needed. Does this memory need to be explicitly accessed by unprivileged code (user mode)?

MaureenHelm replied:
Indeed it does consume an MPU region, which is why I didn't initially set the dts chosen node by default for all i.MX RT boards. I mistakenly thought I needed the MPU region for any execute access. I don't currently see a need for user access, so the MPU region has now been removed.

ioannisg replied:
very well

MaureenHelm committed:
boards: soc: arm: Set zephyr,itcm chosen node for i.MX RT boards

Sets the device tree chosen node for instruction tightly coupled memory
(ITCM) on all i.MX RT boards. Leverages the common Cortex-M linker
section instead of the SoC-specific one.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm commented:
@ioannisg will you please have another look?
